src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -2526,16 +2527,55 @@ namespace winrt::Microsoft::Terminal::Control::implementation
 
             if (items.Size() > 0)
             {
-                std::wstring allPaths;
-                for (auto item : items)
+                std::vector<std::wstring> fullPaths;
+
+                // GH#14628: Workaround for GetStorageItemsAsync() only returning 16 items
+                // at most when dragging and dropping from archives (zip, 7z, rar, etc.)
+                if (items.Size() == 16 && e.DataView().Contains(winrt::hstring{ L"FileDrop" }))
                 {
-                    // Join the paths with spaces
-                    if (!allPaths.empty())
+                    auto fileDropData = co_await e.DataView().GetDataAsync(winrt::hstring{ L"FileDrop" });
+                    if (fileDropData != nullptr)
+                    {
+                        auto stream = fileDropData.as<IRandomAccessStream>();
+                        stream.Seek(0);
+
+                        uint32_t streamSize = (uint32_t)stream.Size();
+                        Buffer buf(streamSize);
+                        auto buffer = co_await stream.ReadAsync(buf, streamSize, InputStreamOptions::None);
+
+                        const HGLOBAL hGlobal = buffer.data();

```

Comment:
Does this need to be wrapped up in some wort of `wil` smart object to manage its lifetime? Kinda seems like we'd just leak it here, but I don't know who owns it's lifetime.

Reply:
I thought that too, but it turns out that `IBuffer.data()` is a `cppwinrt` extension method [implemented directly as](https://github.com/microsoft/cppwinrt/blob/a2131900b9dc9a317ea041abc2927656246089dd/cppwinrt/code_writers.h#L1307):
```cpp
auto data() const
{
    uint8_t* data{};
    static_cast<D const&>(*this).template as<IBufferByteAccess>()->Buffer(&data);
    return data;
}
```
Meaning that the lifetime of the pointer ["is controlled by the lifetime of the buffer object."](https://learn.microsoft.com/en-us/cpp/cppcx/obtaining-pointers-to-data-buffers-c-cx?view=msvc-170#getpointertopixeldata)
With the way we initialized the buffer, it has a scoped lifetime which means we should be safe from leaks.

Now, I admit I have no idea how `Buffer` is implemented internally (closed-source...) so I can't say whether having it stack allocated or heap allocated. I imagine that the data is marshalled through and we are really only storing a pointer (COM pointer?) on the stack as it is.

Reply:
This seems totally safe; thanks for the explanation!